src/D2L.CodeStyle.Analyzers/Immutability/ImmutabilityAnalyzer.ImmutableTypeParameterArgumentAnalysis.cs

```diff
@@ -0,0 +1,654 @@
+
+				AnalyzeTypeRecursive(
+					reportDiagnostic,
+					annotationsContext,
+					immutabilityContext,
+					parameter.Type,
+					() => getParameterSyntax( parameter.Ordinal )
+				);
+			}
+		}
+
+		private static void AnalyzeMethodUsage(
+			OperationAnalysisContext ctx,
+			AnnotationsContext annotationsContext,
+			ImmutabilityContext immutabilityContext,
+			IMethodSymbol method,
+			Func<SyntaxNodeOrToken> getAnalyzedSyntax
+		) {
+			AnalyzeTypeArgumentsRecursive(
+				ctx.ReportDiagnostic,
+				annotationsContext,
+				immutabilityContext,
+				method.TypeArguments,
+				method.TypeParameters,
+				getAnalyzedSyntax
+			);
+
+			AnalyzeMaybeStaticMemberAccess(
+				ctx,
+				annotationsContext,
+				immutabilityContext,
+				method,
+				getAnalyzedSyntax
+			);
+		}
+
+		private static void AnalyzeMaybeStaticMemberAccess(
+			OperationAnalysisContext ctx,
+			AnnotationsContext annotationsContext,
+			ImmutabilityContext immutabilityContext,
+			ISymbol member,
+			Func<SyntaxNodeOrToken> getAnalyzedSyntax
+		) {
+			if( !member.IsStatic ) {
+				return;
+			}
+
+			AnalyzeTypeRecursive(
+				ctx.ReportDiagnostic,
+				annotationsContext,
+				immutabilityContext,
+				member.ContainingType,
+				SelectLeftSyntax( getAnalyzedSyntax )
+			);
+		}
+
+		private static void AnalyzeTypeRecursive(
+			Action<Diagnostic> reportDiagnostic,
+			AnnotationsContext annotationsContext,
+			ImmutabilityContext immutabilityContext,
+			ITypeSymbol type,
+			Func<SyntaxNodeOrToken> getAnalyzedSyntax
+		) {
+			INamedTypeSymbol? namedType = GetNamedTypeRecursive( type );
+			if( namedType is null ) {
+				return;
+			}
+
+			if( namedType.ContainingType is not null ) {
+				AnalyzeTypeRecursive(
+					reportDiagnostic,
+					annotationsContext,
+					immutabilityContext,
+					namedType.ContainingType,
+					SelectLeftSyntax( getAnalyzedSyntax )
+				);
+			}
+
+			AnalyzeTypeArgumentsRecursive(
+				reportDiagnostic,
+				annotationsContext,
+				immutabilityContext,
+				namedType.TypeArguments,
+				namedType.TypeParameters,
+				getAnalyzedSyntax
+			);
+
+			static INamedTypeSymbol? GetNamedTypeRecursive( ITypeSymbol type ) => type switch {
+				INamedTypeSymbol namedType => namedType,
+				IArrayTypeSymbol arrayType => GetNamedTypeRecursive( arrayType.ElementType ),
+				_ => null
+			};
+		}
+
+		private static void AnalyzeTypeArgumentsRecursive(
+			Action<Diagnostic> reportDiagnostic,
+			AnnotationsContext annotationsContext,
+			ImmutabilityContext immutabilityContext,
+			ImmutableArray<ITypeSymbol> typeArguments,
+			ImmutableArray<ITypeParameterSymbol> typeParameters,
+			Func<SyntaxNodeOrToken> getAnalyzedSyntax
+		) {
+			getAnalyzedSyntax = SelectRightSyntax( getAnalyzedSyntax );
+
+			foreach( ITypeParameterSymbol typeParameter in typeParameters ) {
+				ITypeSymbol argument = typeArguments[ typeParameter.Ordinal ];
+
+				SyntaxNodeOrToken getThisArgument() => GetTypeArgumentSyntax( getAnalyzedSyntax(), typeParameter.Ordinal );

```

Comment:
The recursive closure wrapping stuff was pretty confusing. I don't have a concrete suggestion otherwise, just a note.